src/Interpreters/Access/InterpreterCreateUserQuery.cpp

```diff
@@ -2,31 +2,143 @@
 #include <Parsers/Access/ASTCreateUserQuery.h>
 #include <Parsers/Access/ASTRolesOrUsersSet.h>
 #include <Parsers/Access/ASTUserNameWithHost.h>
+#include <Parsers/Access/ASTAuthenticationData.h>
 #include <Parsers/ASTDatabaseOrNone.h>
 #include <Access/AccessControl.h>
 #include <Access/ContextAccess.h>
 #include <Access/User.h>
 #include <Interpreters/Access/InterpreterSetRoleQuery.h>
 #include <Interpreters/Context.h>
 #include <Interpreters/executeDDLQueryOnCluster.h>
+#include <Storages/checkAndGetLiteralArgument.h>
 #include <boost/range/algorithm/copy.hpp>
 
+#include "config.h"
+
+#if USE_SSL
+#     include <openssl/crypto.h>
+#     include <openssl/rand.h>
+#     include <openssl/err.h>
+#endif
 
 namespace DB
 {
 namespace ErrorCodes
 {
     extern const int BAD_ARGUMENTS;
+    extern const int SUPPORT_IS_DISABLED;
+    extern const int OPENSSL_ERROR;
 }
 namespace
 {
+    AuthenticationData makeAuthenticationData(const ASTAuthenticationData & query, ContextPtr context, bool check_password_rules)
```

Comment:
Could we move that transformation function to Access/AuthenticationData.h/cpp?

Reply:
It sounds like a good idea, but I'm afraid that it will run into the same problem, `clickhouse_common_access` is linked to `clickhouse_parsers`.

Reply:
`AuthenticationData.cpp` was in `clickhouse_common_access` only because it was used in `ASTCreateUser`. Now that's not the case anymore.

Reply:
We can do something like this:

``` cpp
static AuthenticationData::fromAST(
    ASTPtr query,
    const ASTs & prepared_arguments,
    std::optional<AuthenticationType> default_type
    //, std::optional<int> bcrypt_param - in the future
)
```
InterpreterCreateUserQuery::execute():
``` cpp
...
std::optional<AuthenticationData> auth_data;
if (query.auth_data)
{
    if (!query.attach)
    {
        auto password = auth_data->getPassword();
        if (password)
            global_context->getAccessControl().checkPasswordComplexityRules(*password);
    }

    AuthenticationType default_type = access_control.getDefaultPasswordType();
    ASTs args(args_size);
        for (size_t i = 0; i < args_size; ++i)
            args[i] = evaluateConstantExpressionAsLiteral(query.children[i], context);

    auth_data = AuthenticationData::fromAST(query, args, default_type);
}
...
```

Is it better?

Reply:
> But `AuthenticationData.h/cpp` is used both in `ASTAuthenticationData` and in `ParseCreateUserQuery` because of the `AuthenticationType`.

We can keep the `AuthenticationType` enum in the folder `src/Access/Common` and move `AuthenticationData` to `src/Access`.